docs/design/service-modules.md

```diff
@@ -0,0 +1,233 @@
+be defined in the module info by using the `uses` keyword.
+
+Based on this given constraints the `module-info.java` of an api module looks like this:
+
+```
+module com.hedera.node.app.service.foo {
+    exports com.hedera.node.app.service.foo;
+    
+    uses com.hedera.node.app.service.foo.FooService;
+    
+    requires transitive com.hedera.node.app.spi;
+}
+```
+
+Next to this the `build.gradle.kts` file should look like this:
+
+```
+plugins {
+    id("com.hedera.hashgraph.conventions")
+}
+
+description = "Hedera Foo Service API"
+
+dependencies {
+    api(project(":hedera-node:hedera-app-spi"))
+}
+```
+
+The public service api can depend on additional libraries.
+Such libraries will be defined as **public api** and must be added as `api(...)` dependencies to gradle and defined
+as `requires transitive` in the `module-info.java`.
+No api module must ever depend on any service implementation or the `hedera-mono-service` module.
+The `@NonNull`/`@Nullable` annotations can be used in the service definitions.
+Since the annotations are already defined as transitive dependencies at compile time by the `hedera-app-spi` module
+no extra dependency needs to be added.
+Unit tests and integration tests are normally part of the implementation modules but the api modules can contain
+such test when it makes sense. In such case the same definition as for the implementation modules should be used.
+
+Since all services must be accessed by the Java SPI we need to define an entry-point in the api module.
+Here we use a pattern that defines the entry point directly as a static method in the service interface:
+
+```
+package com.hedera.node.app.service.foo;
+
+import com.hedera.node.app.spi.Service;
+import com.hedera.node.app.spi.ServiceFactory;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.util.ServiceLoader;
+
+public interface FooService extends Service {
+
+    @NonNull
+    @Override
+    default String getServiceName() {
+        return FooService.class.getSimpleName();
+    }
+
+    /**
+     * Returns the concrete implementation instance of the service
+     *
+     * @return the implementation instance
+     */
+    @NonNull
+    static FooService getInstance() {
+        return ServiceFactory.loadService(FooService.class, ServiceLoader.load(FooService.class));
+    }
+}
+```
+
+The basic `Service` interface and the `ServiceFactory` class that provides the SPI implementation are part of
+the `hedera-app-spi` module.
+
+## service implementation modules
+
+For each api module an additional module that contains the implementation must be provided.
+The structure of a minimal implementation module (imp-module) looks like this:
+
+```
+hedera-foo-service-impl/
+├── src/main/java/
+│   ├── com.hedera.node.app.service.foo.impl
+│   │   ├── FooServiceImpl.java
+│   │   └── package-info.java
+│   └── module-info.java
+├── src/itest/java/
+│   ├── com.hedera.node.app.service.foo.impl.itest
+│   │   ├── FooServiceImplITest.java
```

Comment:
Is this the naming pattern for integration tests, we agreed upon? I am more familiar with `*IT.java`.

Reply:
Good Question. I have no idea

Reply:
Asked in Slack about it